File: python/121_Best_Time_to_Buy_and_Sell_Stock.py
# One pass. O(n) time, O(1) space

class Solution(object):
    def maxProfit(self, prices):
        """
        :type prices: List[int]
        :rtype: int
        """
        minprice = float("inf")
        maxprofit = 0
        
        for currprice in prices:
            if currprice < minprice:
                minprice = currprice 
            maxprofit = max(maxprofit, currprice-minprice)
        return maxprofit

# DP is also ok. O(n) time, O(n) space

File: java/121_Best_Time_to_Buy_and_Sell_Stock.java
class Solution {
    public int maxProfit(int[] prices) {
        int min_price = Integer.MAX_VALUE;
        int max_profit = 0;
        for (int i = 0; i < prices.length; i++) {
            if (prices[i] < min_price) min_price = prices[i];
            else if (prices[i] - min_price > max_profit) {
                max_profit = prices[i] - min_price;
            } 
        }
        return max_profit;
    }
}

class Solution {
    public int maxProfit(int[] prices) {
        int buy = 0, res = 0;
        for (int i = 0; i < prices.length; i++) {
            if (prices[buy] < prices[i]) {
                res = Math.max(res, prices[i] - prices[buy]);
            }
            else {
                buy = i;
            }
        }
        return res;
    }
}